shared-worker/firebase-worker.js

```
@@ -0,0 +1,54 @@
+importScripts('../firebase/firebase-app.js');
+importScripts('../firebase/firebase-firestore.js');
+var config = {
+  // your config here
+};
+
+/** Port used to communicate with the shared worker. */
+self.sharedWorkerPort = null;
+
+/** Whether this Firestore worker holds the lease. */
+self.hasPrimaryLease = false;
+
+/** The Firestore instance. */
+self.db = null;
+
+self.onmessage = function(e) {
+  // Setup message listeners after receiving the Shared Worker's port and
+  // request the lease.
+  if (e.data.sharedWorkerPort) {
+    self.sharedWorkerPort = e.data.sharedWorkerPort;
+    self.sharedWorkerPort.onmessage = (event) => handleLeaseEvent(event);
+    self.sharedWorkerPort.postMessage('requestingLease');
+  } 
+
+  // Release the primary lease in the shared worker if this worker owns the
+  // lease when this worker is being terminated.
+  if (e.data === 'terminate' && self.hasPrimaryLease) {
+    self.sharedWorkerPort.postMessage('releaseLease');
+  } 
+};
+
+async function initializeApp(takeLease) {
+  const app2 = firebase.initializeApp(config, "some-namez");
+  self.db = firebase.firestore(app2);
+  if (takeLease) {
+    await self.db.enablePersistence({experimentalForce: true});
+  } else {
+    await self.db.enablePersistence();
+  } 
+}
+
+function handleLeaseEvent(e) {
+  if (e.data === 'leaseAvailable') {
+    // Start Firestore as the primary lease holder.
+    self.sharedWorkerPort.postMessage('holdLease');
+    self.hasPrimaryLease = true;
+    console.log('Starting Firestore as lease holder.');
+    initializeApp(true);
```

Looks like we're silently ignoring the Promise results?